Aptoide/aptoide-client-v8
# Bugfix/an 1848 timeline adapter leak

What does this PR do?

   This pull-request fixes the timeline cards adapter leak, that was being held in memory with the fragment and not being destroyed when the fragment's view was destroying, leaking every view associated to it.

   The adapter was bound to the Fragments life-cycle instead of view's because it is a requirement to go back to the same state of the list (position & card) when user navigates into an app from a Timeline and presses back (view would get destroyed and created, reason why I associated the adapter to the Fragments life-cycle, holding it in memory).  This would be very bad for memory specially because Timelines can and will be stacked if user navigates to other users Timelines, possibly leading to an OutOfMemoryException.
   
   The solution for this was 3 changes: First clean the adapter when **TimelineFragment**'s view is destroyed. This would break the requirement above explained, making the user go to top everytime he would go back to Timeline from an app. Therefor a second change was submitted: There is now a **TimelinePostsRepository** which has 2 different DataSources (inMemory - cache & remote): The cache is a simple list of posts that will give us the last used list cards and the Remote source is for the first call (when cache is clean) or when user pulls-to-refresh, "invalidating" the in memory cache". The 3rd change was handling list states on **TimelineFragment**'s life-cycle with the savedInstanceState. This would allow to handle configuration changes aswell - which are now finally being taken care!

   Because there are 2 types of Timeline (home & userId timeline) we had to create a mechanism that would cache different types of Timeline and return the right Timeline to the **TimelineFragment**, so now there is a **TimelineRepositoryFactory** which is in the Application scope that returns a **TimelinePostsRepository** (the one with cache+remote). This would make possible to cache different Timeline post lists during a session.

   

Where should the reviewer start?

- [ ] TimelinePostsRepository (memory + remote post sources)
- [ ] TimelineFragment (adapter + list states + on configuration changes)
- [ ] TimelineRepositoryFactory

How should this be manually tested?

  Open Aptoide v8 > Timeline > scroll down a bit > navigate to an App card (get app) > press back. Should return to the same position with the same card showing up. 
  Open Aptoide v8 > Timeline > scroll down a bit > On screen orientation change should now make you stay in the same spot.
  Open Aptoide v8 > Timeline > scroll down a bit > Go to Stores tab > ana2016 store > Timeline (should be different from home Timeline) > scroll down a bit > go to the Games tab inside ana's store > go back to Timeline > should be in the same spot.

What are the relevant tickets?

Tickets related to this pull-request: [AN-1848](https://aptoide.atlassian.net/browse/AN-1848)

Screenshots (if appropriate):
N/A

Questions:

   Is there a blog post? N/A
   Does this add new dependencies which need to be added? No
